trio/_threads.py

```diff
@@ -369,21 +285,30 @@ def do_release_then_return_result():
 
     # This is the function that runs in the worker thread to do the actual
     # work and then schedule the call to report_back_in_trio_thread_fn
-    def worker_thread_fn():
+    # Since this is spawned in a new thread, the trio token needs to be passed
+    # explicitly to it so it can inject it into thread local storage
+    def worker_thread_fn(trio_token):
+        TOKEN_LOCAL.token = trio_token
         result = outcome.capture(sync_fn, *args)
         try:
             token.run_sync_soon(report_back_in_trio_thread_fn, result)
         except trio.RunFinishedError:
             # The entire run finished, so our particular task is certainly
             # long gone -- it must have cancelled.
             pass
+        finally:
+            del TOKEN_LOCAL.token
```

Comment:
This is a finicky style comment; I think the code you have here works correctly. But the `try` block is doing two logically unrelated things here – cleaning up the thread local, and dealing with a missing trio thread. And the thread local cleanup logically has a larger scope – if `capture` raised an exception, we'd still want to clean up the thread local. Fortunately `capture` can't raise an exception, so we could get away with this, but for readability, I think it will be clearer what's happening if we split it into two try blocks, structured like:

```python
TOKEN_LOCAL = ...
try:
    result = ...
    try:
        token.run_sync_soon(...)
    except RunFinishedError:
        ...
finally:
    ...
```

Reply:
Hmm that is an interesting way to look at it, I never thought about the separation of concerns that way!